docs(blog): add testing http requests post (#2282)
refinedev/refine: documentation/blog/2022-08-04-testing-http-requests.md

---
title: Testing HTTP Requests in React Using Nock
description: How to test API calls in React?
slug: test-http-requests-in-react
authors: necati
tags: [react, testing, nock, jest, mocking]
image: https://refine.dev/img/refine_social.png
hide_table_of_contents: false
---

import testrun from '@site/static/img/blog/2022-08-04-testing-http-requests/test-run.png';
import nockrecord from '@site/static/img/blog/2022-08-04-testing-http-requests/nock-recorder-log.png';

## Introduction

> Writing unit tests is very important for the development process. Testing components that use HTTP requests sometimes may be a real pain.
> 
> In testing, we often want to make mock requests in order to test our code without actually making a real HTTP request. This can be especially important when we are testing code that makes external API calls, since we don't want to rely on the availability of the external API.
> 
> We’ll use a third-party package called [nock](https://github.com/nock/nock) that helps us to mock HTTP requests. With nock, we can specify the desired behavior of our mock HTTP requests, including the URL, headers, and body. This allows us to test our code against a known set of data, which can make debugging and testing much easier.

Reviewer: ```suggestion
Writing unit tests is very important for the development process. Testing components that use HTTP requests sometimes may be a real pain.

In testing, we often want to make mock requests to test our code without actually making an HTTP request. This can be especially important when we are testing code that makes external API calls since we don't want to rely on the availability of the external API.

We'll use a third-party package called [nock](https://github.com/nock/nock) that helps us to mock HTTP requests. With nock, we can specify the desired behavior of our mock HTTP requests, including the URL, headers, and body. This allows us to test our code against a known data set, making debugging and testing much more straightforward.
```

<!--truncate-->

I'll show how to write unit tests for API calls by mocking method in the simple React app.

Steps we'll cover: 

- [Introduction](#introduction)
- [Why mocking HTTP requests during testing is important?](#why-mocking-http-requests-during-testing-is-important)
- [What is Nock?](#what-is-nock)
- [Bootstrapping the example app](#bootstrapping-the-example-app)
- [Adding a unit test](#adding-a-unit-test)
- [Nock installation and configuration](#nock-installation-and-configuration)
- [Custom requests in Nock](#custom-requests-in-nock)
    - [All HTTP methods like `GET`, `POST`, `PUT`, `DELETE` can be mock.](#all-http-methods-like-get-post-put-delete-can-be-mock)
    - [To handle query parameters, the `query` option can be used.](#to-handle-query-parameters-the-query-option-can-be-used)
    - [Mocking server Errors](#mocking-server-errors)
- [Recording in Nock](#recording-in-nock)
- [Alternative API mocking libraries](#alternative-api-mocking-libraries)
- [Conclusion](#conclusion)

## Why mocking HTTP requests during testing is important?

> Mock testing is a great way to speed up the process of running tests because you can eliminate external systems and servers.
> 
> 
> These are all possible errors that you might encounter when running tests with the API:
> 
>   - The data returned from API can be different on each request.
>   - It takes a longer time to finish running the test.
>   - You may get a big size of data that you don't need to use in tests.
>   - You may have issues like rate limiting and connectivity.
> 
> 
> We'll use the Nock to find a solution for these problems. We'll create a simple react app and make a request to an external API. We are going to implement how to mock API calls and write a unit test for API calls using Nock in a React application.

Reviewer: ```suggestion
Mock testing is a great way to speed up running tests because you can eliminate external systems and servers.

These are all possible errors that you might encounter when running tests with the API:

  - The data returned from API can be different on each request.
  - It takes a longer time to finish running the test.
  - You may get a big size of data that you don't need to use in tests.
  - You may have issues like rate limiting and connectivity.

We'll use the Nock to find a solution for these problems. We'll create a simple react app and request an external API. We will implement how to mock API calls and write a unit test for API calls using Nock in a React application.
```

## What is Nock?

> Nock is an HTTP server mocking and expectations library. Nock works by overriding Node's `http.request` function.
> 
> 
> It helps us mock calls to API and specifies what URLs we want to listen for and responds with predefined responses, just like how real APIs would do it.
> 
> We can use nock to test React components that make HTTP requests.

Reviewer: ```suggestion
Nock is an HTTP server mocking and expectations library. Nock works by overriding Node's `http.request` function.

It helps us mock calls to API and specifies what URLs we want to listen for, and responds with predefined responses, just like real APIs would.

We can use nock to test React components that make HTTP requests.
```
